### ISSUE TITLE
Support single overload for functions

### ISSUE THREAD

**Issue by mezuzza** [microsoft/pyright#4242]
**Describe the solution you'd like**
Currently, there is no way to specify the allowed keywords for a `**kwargs` afaict. `ParamSpec` almost does this, but only for higher order functions that don't modify kwargs. There was [an attempt][kwarg-concat] to solve this, but it was ultimately deferred.

Unfortunately, python's `kwargs` is not equivalent to specifying default values as there's no way to tell in the latter case when a user explicitly set a value.

[kwarg-concat]: https://peps.python.org/pep-0612/#concatenating-keyword-parameters

As a solution to this, I've taken to using `@overload` as a means of specifying kwargs.

As an example:
```python
@dataclass
class Bar:
  a: str = field(default_factory="this is set if you don't pass a to foo")
  b: int = 42
  c: float = 3.14

@overload
def foo( # type: ignore
  a: str|None=None, b: int|None=None, c: float|None=None):
   ...

def foo(**kwargs) -> Bar:
  return Bar(**{k: v for k, v in kwargs.items() if v is not None})
```

The only problem is that I need a `#type: ignore` comment at the moment as pyright is wondering where the other overloads are. It's a reasonable question, but could we drop that warning? It doesn't feel like a heavy lift and it seems like a valid use case.

If there's some other solution I'm missing, let me know.

**Additional context**
Python: 3.10.8
Pyright: 1.1.280

**Comment by erictraut** [comment]
This isn't an appropriate use for `@overload`. 

`@overload` is intended for functions that are polymorphic — i.e. their return type depends on the input signature. The function in your code sample is not polymorphic. It has a single, well-defined signature and a single return type.

The proper way to annotate this function is:

```python
def foo(a: str|None=None, b: int|None=None, c: float|None=None) -> Bar:
  return Bar(**{k: v for k, v in {"a": a, "b": b, "c": c} if v is not None})
```

**Comment by mezuzza** [comment]
I don't really see how this isn't an appropriate use of overload. I understand what problem it's intended to solve, but the solution can be used to solve more than one problem.

I apologize that my example wasn't quite great at providing motivation for this. I've updated it below. There's no way to annotate the function to say "You can provide exactly the arguments that `Bar` accepts, and if you do provide them, I'll use them instead of `Bar`'s defaults". 

The problem with any default argument approach including the suggestion you had is that there's no way to discriminate whether the user passed in a variable explicitly or not. Using kwargs, you can. 

```python
@dataclass
class Bar:
  a: str | None = field(default_factory="this is set if you don't pass a to foo")
  b: int | None = 42
  c: float | None = None

@overload
def foo( # type: ignore
  a: str|None=None, b: int|None=None, c: float|None=None):
   ...

def foo(**kwargs) -> Bar:
  return Bar(**kwargs)
 ```

**Comment by hmc-cs-mdrissi** [comment]
You could use https://discuss.python.org/t/pep-692-using-typeddict-for-more-precise-kwargs-typing/17314/51 potentially. That allows using kwargs while specifying the exact expected arguments.

**Comment by mezuzza** [comment]
I would love to use pep-692, but unfortunately it's still a draft and won't be available until 3.12. 

As for the latter, yes, you're right I could use that. Not my favorite solution, but definitely possible. Honestly, this single overload pattern is more similar to 692 than introducing a private class. Still, I'll try to use this pattern moving forward if necessary.

That said, I still don't see why `overload` should only be used for 2+ forms. It naturally works for 1. It's like we we're doing induction and we just chose a base case of 2 instead of 1 even though it works just fine.

What I've heard so far is an argument for why things are the way they are - which is totally reasonable by the way. But I what I haven't heard (or maybe don't appreciate) is why we shouldn't improve this and permit a single overload case? How would it be abused?

That said, I won't push after this. Personally, I think the convincing argument for me is:
1. There are at least two workarounds (use `#type: ignore` or use a private singleton)
2. We have a better solution on the horizon which invalidates this. So while it sucks to use workarounds for now, we won't be using them for long.
3. All that leaves is some argument for elegance of having overloads start at 1 form, but it doesn't really solve a problem at that point, so who cares. No reason to do something just to make it prettier.

**Comment by erictraut** [comment]
@mezuzza, you are using the wrong tool for the job. It's as if you're using a screwdriver to pound in a nail. It kind of works, but it's definitely not what a screwdriver was designed for. 

The error that you see when a single overload is used is there for a reason. Users misunderstand how overloads are meant to be used, and this error guides them to the correct solution. Removing the error might suit your needs, but it's not a good idea in general.

If you like the PEP 692 solution, you don't need to wait for Python 3.12. It's implemented already in pyright. You will need to use the `type_extensions.Unpack` symbol rather than the more elegant `**` syntax since the latter requires changes in the Python parser. See the [Backward Compatibility](https://peps.python.org/pep-0692/#backwards-compatibility) section for details. The only risk is that PEP 692 could be rejected by the steering council, and then I would remove support for it in a future version of pyright.
